### PR TITLE
Multi collateral staging branch

### PR DESCRIPTION
**Pull request by AntonAndell**

### CODE REVIEW

File: core-contracts/BalancedOracle/src/main/java/network/balanced/score/core/balancedoracle/BalancedOracleImpl.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        dexPricedAssets.set("BALN",  BigInteger.valueOf(3));
+    }
+    
+    @External(readonly = true)
+    public String name() {
+        return TAG;
+    }
+
+    @External
+    public BigInteger getPriceInLoop(String symbol) {
```
Comment (AntonAndell) [inline core-contracts/BalancedOracle/src/main/java/network/balanced/score/core/balancedoracle/BalancedOracleImpl.java:60]: Add update time limits, perhaps discuss with Dan

File: core-contracts/Dex/src/test/java/network/balanced/score/core/dex/DexTestCore.java [hunk 2]
```diff
@@ -615,263 +501,6 @@ void getTotalValue() {
...
-        prep._votes_in_per = BigInteger.valueOf(100);
-        PrepDelegations[] preps = new PrepDelegations[]{prep};
-
-        contextMock.when(() -> Context.call(eq(stakingScore.getAddress()), eq("delegate"), any())).thenReturn(
-                "Staking delegate called");
-        dexScore.invoke(governanceScore, "delegate", (Object) preps);
-
-        contextMock.verify(() -> Context.call(eq(stakingScore.getAddress()), eq("delegate"), any()));
-    }
-
```
Comment (AntonAndell) [inline core-contracts/Dex/src/test/java/network/balanced/score/core/dex/DexTestCore.java:874]: do not remove

File: core-contracts/Governance/src/intTest/java/network/balanced/score/core/governance/GovernanceIntegrationTest.java [hunk 3]
```diff
@@ -58,6 +58,7 @@ static void setup() throws Exception {
         tester.rewards.claimRewards();
 
         BigInteger balance = tester.baln.balanceOf(tester.getAddress());
+        System.out.println(balance);
```
Comment (AntonAndell) [inline core-contracts/Governance/src/intTest/java/network/balanced/score/core/governance/GovernanceIntegrationTest.java:61]: remove

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java [hunk 4]
```diff
@@ -495,55 +301,62 @@ public void retireBadDebt(String _symbol, BigInteger _value) {
...
-        checkDistributions(day, newDay);
+   
+        BigInteger totalBadDebt = BigInteger.ZERO;
+        BigInteger remaningValue = _value;
+        for (String collateralSymbol :  CollateralDB.getCollateral().keySet()) {
+            BigInteger badDebt = asset.getBadDebt(collateralSymbol);
+            BigInteger badDebtAmount = badDebt.min(remaningValue);
+
+            BigInteger collateralToRedeem = badDebtRedeem(from, collateralSymbol, asset, badDebtAmount);
+            transferCollateral(collateralSymbol, from, collateralToRedeem, "Bad Debt redeemed.", new byte[0]);
```
Comment (AntonAndell) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java:318]: Move transfer to the end

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java [hunk 5]
```diff
@@ -554,321 +367,302 @@ public void returnAsset(String _symbol, BigInteger _value, @Optional boolean _re
...
+        transferCollateral(collateralSymbol, Context.getCaller(), reward, "Liquidation reward of", new byte[0]);
         AssetDB.updateDeadMarkets();
 
-        if (isBeforeContinuousRewardDay) {
-            PositionsDB.removeNonZero(position.getId());
-        }
-
         String logMessage = collateral + " liquidated from " + _owner;
-        Liquidate(_owner, collateral, logMessage);
+        liquidated(_owner, collateral, logMessage);
```
Comment (AntonAndell) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java:606]:
```suggestion
        Liquidated(_owner, collateral, logMessage);
```

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java [hunk 6]
```diff
@@ -1159,11 +986,11 @@ public void LoanRepaid(Address account, String symbol, BigInteger amount, String
     }
 
     @EventLog(indexed = 3)
-    public void BadDebtRetired(Address account, String symbol, BigInteger amount, BigInteger sicx_received) {
+    public void BadDebtRetired(Address account, String symbol, BigInteger amount) {
     }
 
     @EventLog(indexed = 2)
-    public void Liquidate(Address account, BigInteger amount, String note) {
+    public void liquidated(Address account, BigInteger amount, String note) {
```
Comment (AntonAndell) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/LoansImpl.java:993]:
```suggestion
    public void Liquidated(Address account, BigInteger amount, String note) {
```

File: core-contracts/Rebalancing/src/main/java/network/balanced/score/core/rebalancing/RebalancingImpl.java [hunk 7]
```diff
@@ -150,28 +164,31 @@ public BigInteger getPriceChangeThreshold() {
      *
      * @return {List<Object> }   [<Positive difference>, <Tokens to sell>, <Negative difference>]
      */
+
     @External(readonly = true)
     @SuppressWarnings("unchecked")
-    public List<Object> getRebalancingStatus() {
+    public List<Object> getRebalancingStatusFor(Address address) {
```
Comment (AntonAndell) [inline core-contracts/Rebalancing/src/main/java/network/balanced/score/core/rebalancing/RebalancingImpl.java:170]:
```suggestion
    public List<Object> getRebalancingStatusFor(Address collateralAddress) {
```

File: score-client/src/main/java/foundation/icon/score/client/DefaultScoreClient.java [hunk 8]
```diff
@@ -44,7 +44,7 @@
 
 public class DefaultScoreClient extends JsonrpcClient {
     public static final Address ZERO_ADDRESS = new Address("cx0000000000000000000000000000000000000000");
-    public static final BigInteger DEFAULT_STEP_LIMIT = new BigInteger("9502f900", 16);
```
Comment (AntonAndell) [inline score-client/src/main/java/foundation/icon/score/client/DefaultScoreClient.java:47]: revert this change

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java [hunk 9]
```diff
@@ -80,333 +75,117 @@ public Address getAddress() {
...
-
-    public BigInteger getAssetPosition(String symbol) {
-        Context.require(arrayDbContains(AssetDB.assetSymbols, symbol), TAG + ": " + symbol + " is not a supported " +
-                "asset on Balanced.");
-
-        if (isBeforeContinuousRewardDay() || !getDataMigrationStatus(symbol)) {
-            return getAssets(lastSnap(), symbol);
+    public BigInteger getTotalDebt(String symbol) {
+        BigInteger totalDebt = this.totalDebt.at(dbKey).get(symbol);
+        if (totalDebt == null) {
```
Comment (nightowl121) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:92]: Won't there be any issue if we return total debt of sicx for any non existing token symbol?

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java [hunk 10]
```diff
@@ -30,25 +30,20 @@
...
-            BigInteger.class);
-    private final BranchDB<String, DictDB<String, Boolean>> dataMigrationStatus = Context.newBranchDB("data_migration " +
-            "_status", Boolean.class);
-    private final BranchDB<String, BranchDB<String, DictDB<String, BigInteger>>> loansPosition = Context.newBranchDB(
+    private final BranchDB<String, BranchDB<String, DictDB<String, BigInteger>>> debt = Context.newBranchDB(
             "loan_balance", BigInteger.class);
-    private final BranchDB<String, DictDB<String, BigInteger>> collateralPosition = Context.newBranchDB("collateral_balance"
+    private final BranchDB<String, DictDB<String, BigInteger>> collateral = Context.newBranchDB("collateral_balance"
             , BigInteger.class);
+    private final BranchDB<String, DictDB<String, BigInteger>> totalDebt = Context.newBranchDB(
```
Comment (nightowl121) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:45]: Could you document what do the keys represent in debt and totalDebt?

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java [hunk 11]
```diff
@@ -421,96 +200,83 @@ public Standing getStanding(Integer day, Boolean readOnly) {
...
-                    standing.standing = Standings.MINING;
-                }
-            } else if (standing.ratio.compareTo(lockingRatio.get().multiply(EXA).divide(POINTS)) > 0) {
-                standing.standing = Standings.NOT_MINING;
-            } else if (standing.ratio.compareTo(liquidationRatio.get().multiply(EXA).divide(POINTS)) > 0) {
-                standing.standing = Standings.LOCKED;
-            } else {
-                standing.standing = Standings.LIQUIDATE;
-            }
+        if (standing.ratio.compareTo(liquidationRatio.get(collateralSymbol).multiply(EXA).divide(POINTS)) > 0) {
```
Comment (nightowl121) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:203]: Does this make an assumption that all collateral would have same liquidation ratio?

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/utils/Token.java [hunk 12]
```diff
@@ -47,11 +49,11 @@ public String getPeg() {
     }
 
     public BigInteger priceInLoop() {
-        return (BigInteger) call(tokenAddress, "priceInLoop");
+        return (BigInteger) call(LoansVariables.oracle.get(), "getPriceInLoop", symbol());
```
Comment (nightowl121) [inline core-contracts/Loans/src/main/java/network/balanced/score/core/loans/utils/Token.java:52]: NoteForFYI: After deployment of this update it will require oracle address to be set.

File: core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+        assertEquals(expectedTotalDebt, getTotalDebt());
+        assertEquals(BigInteger.ZERO, loanTakerFullRepayBaS.get("_balance"));
+        assertEquals(debt.divide(BigInteger.TWO), loanTakerPartialRepayBaS.get("_balance"));
+        assertEquals(BigInteger.ZERO, loantakerIETHFullRepayBaS.get("_balance"));
+        assertEquals(debt, loantakerMultiPartialRepayBaS.get("_balance"));
+    }
+
+    @Test
+    @Order(3)
+    void withdrawCollateral() throws Exception {
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTest.java:185]: Yeah mabye, but it is hard to keep structure and tests readable when one tests are not somewhat isolated

File: core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTestBase.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package network.balanced.score.core.loans;
+
+import network.balanced.score.lib.test.integration.Balanced;
+import org.junit.jupiter.api.BeforeAll;
+
+import java.math.BigInteger;
+
+class LoansIntegrationTestBase extends LoansIntegrationTest {
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTestBase.java:24]: This is setup of a fresh Loans test while the other is a migration, However we setup the contract both run the tests in LoansIntegrationTest. So basically run all test with fresh contract and with a migrated one

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/CollateralDB.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+    public static void migrateToNewDBs() {
+        int totalCollateralCount = collateralList.size();
+        if (collateralAddresses.size() > 0) {
+            return;
+        }
+
+        for (int i = 0; i < totalCollateralCount; i++) {
+            String symbol = collateralList.get(i);
+            Collateral collateral = getCollateral(symbol);
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/CollateralDB.java:49]: Yeah, collateral is not a new list, it already exists on mainnet

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java [hunk 16]
```diff
@@ -30,25 +30,20 @@
...
-
     static final String TAG = "BalancedLoansPositions";
+
     private final BranchDB<String, VarDB<Integer>> id = Context.newBranchDB("id", Integer.class);
     private final BranchDB<String, VarDB<BigInteger>> created = Context.newBranchDB("created", BigInteger.class);
     private final BranchDB<String, VarDB<Address>> address = Context.newBranchDB("address", Address.class);
-    private final BranchDB<String, ArrayDB<Integer>> snaps = Context.newBranchDB("snaps", Integer.class);
-    private final BranchDB<String, BranchDB<Integer, DictDB<String, BigInteger>>> assets = Context.newBranchDB("assets",
-            BigInteger.class);
-    private final BranchDB<String, DictDB<String, Boolean>> dataMigrationStatus = Context.newBranchDB("data_migration " +
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:46]: yeah
https://github.com/balancednetwork/balanced-java-contracts/issues/161

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java (hunk 9 above)
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:92]: Not that i can think of, this is only a migration check. And you can't open position for other assets than bnUSD. But im gonna take a through look through this again

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java (hunk 11 above)
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:203]: Not sure i understand, liquidationRatio is unique per collateral `liquidationRatio.get(collateralSymbol)`

File: core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTest.java:185]: When writing them i had the main focus of them being thorough, but i can see the point in optimizing them.

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/CollateralDB.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/CollateralDB.java:49]: LoansIntegrationTestMigration file does this

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java (hunk 9 above)
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/positions/Position.java:92]: Think of it like this, either a user enters balanced with no debt or with sICX debt. After this migration you can have debt for sICX collateral without increasing your totalDebt across all collaterals

File: core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/Collateral.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import score.Context;
+import score.VarDB;
+import scorex.util.HashMap;
+
+import java.util.Map;
+
+public class Collateral {
+
+    private final BranchDB<String, VarDB<Address>> assetAddress = Context.newBranchDB("address", Address.class);
+    private final BranchDB<String, VarDB<Boolean>> active = Context.newBranchDB("active", Boolean.class);
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/Collateral.java:31]: They are not new, they retain their values from the collateral currently assets on mainnet

[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/Collateral.java:31]: We do thus have a limitation of a collateraltype can't be a assetType, which is a bad idea anyway

[the remark this answers is not in the document]
Comment (nightowl121) [reply core-contracts/Loans/src/main/java/network/balanced/score/core/loans/collateral/Collateral.java:31]: Yeah, it seems they are shared DBs with other classes.

File: core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTestBase.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (AntonAndell) [reply core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTestBase.java:24]: That is an abstract class and does not run any tests LoansIntegrationTestBase runs the tests in LoansIntegrationTest. It is done so we can run the full test suite with and without migration

[the remark this answers is not in the document]
Comment (nightowl121) [reply core-contracts/Loans/src/intTest/java/network/balanced/score/core/loans/LoansIntegrationTestBase.java:24]: Didn't notice that it was abstract class.

**Review by nightowl121 — APPROVED** [review]
